Issue #5394 Fix injection for quickstart

janbartel opened the pull request:
Closes #5394 

Resource injection was broken with jetty-10 and quickstart. We made changes to accord with the servlet tck tests and servlet 4.0 clarifications that if a web descriptor was metadata-complete, we would not do resource injections. The descriptor that quickstart generated was marked as metadata-complete, thus we did not do resource injections.  I think it is incorrect to mark the quickstart-web.xml as metadata-complete: according to the clarified wording of the spec section 8.1, it would only be metadata-complete iff it contained the declarative equivalents of all resource injection annotations (and others like post-construct, pre-destroy etc etc). As jetty "introspects" to find these kind of annotations at the point at which an object is going into service, rather than during the blanket annotation code-scan,  the quickstart-web.xml will never contain them and thus cannot be considered metadata-complete.

So, the solution is to simply mark the generated quickstart-web.xml as metadata-complete=false.

I've also changed one of the quickstart tests to really check properly for injection/post-construct etc happening correctly - previously the test was just testing if the webapp got deployed by retrieving static content, which is not sufficient.

janbartel committed:
Issue #5394 Fix injection for quickstart

Signed-off-by: Jan Bartel <janb@webtide.com>

olamy commented on jetty-quickstart/src/main/java/org/eclipse/jetty/quickstart/PreconfigureQuickStartWar.java:
```diff
@@ -109,7 +110,7 @@ public static void preconfigure(Resource war, Resource dir, Resource xml) throws
         webapp.setAttribute(QuickStartConfiguration.MODE, QuickStartConfiguration.Mode.GENERATE);
         webapp.setAttribute(QuickStartConfiguration.ORIGIN_ATTRIBUTE, "");
-
+        webapp.setAttribute(MetaInfConfiguration.CONTAINER_JAR_PATTERN, ".*/jetty-servlet-api-[^/]*\\.jar$|.*/javax.servlet.jsp.jstl-.*\\.jar$|.*/org.apache.taglibs.taglibs-standard-impl-.*\\.jar$");
```
take care when merging to 11 ;)

gregw commented on jetty-quickstart/src/main/java/org/eclipse/jetty/quickstart/QuickStartGeneratorConfiguration.java:
```diff
@@ -157,7 +157,7 @@ public void generateQuickStartWebXml(WebAppContext context, OutputStream stream)
         webappAttr.put("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance");
         webappAttr.put("xsi:schemaLocation", "http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_" + (major + "_" + minor) + ".xsd");
-        webappAttr.put("metadata-complete", "true");
+        webappAttr.put("metadata-complete", "false");
```
Surely this should not be hard coded to false, but take the value from the main web.xml
Of course there will be issues if some fragments are complete and others are not!!!

janbartel replied:
Ah, good point. I'll fix.

janbartel committed:
Issue #5394 Set quickstart metadatacomplete based on web.xml value

Signed-off-by: Jan Bartel <janb@webtide.com>

janbartel commented:
I've fixed the original problem (and addressed @gregw's concern about forcing the metadata-complete to false for the quickstart-web.xml).

However, this does bring up a wider problem that will need substantially more work to address:  because we lose the origin of the servlet/filter/listener etc once we generate the quickstart-web.xml, if they came from a web-fragment.xml that was metadata-complete, we would incorrectly introspect it. According to the spec, if a servlet/filter/listener came from a metadata-complete fragment, it should not be injected. 

We either need to somehow retain more information about the origin of each servlet/filter/listener when generating quickstart, or quickstart needs to do the introspection (or the equivalent with asm) and generate the equivalent web.xml statements for all injections, post-construct/pre-destroy etc.
